### ISSUE TITLE
Facelet cache wrong expiration threshold

### ISSUE THREAD

**Issue by ren-zhijun-oracle** [eclipse-ee4j/mojarra#4182]
The patch submitted from [https://java.net/jira/browse/JAVASERVERFACES-4107](https://java.net/jira/browse/JAVASERVERFACES-4107) induces the side-effect that reload-check happens too often.
Generally speaking this is a major issue only if your app makes extensive use of composite components, templates and/or fragment includes.
In all other cases the performance impact is negligible.

```
final class DefaultFaceletCache extends FaceletCache<DefaultFacelet> {

	long getAndUpdateNextRefreshTime() {
		// There is no point in calculaing the next refresh time if we are refreshing always/never 		 return (_refreshInterval > 0) ? _nextRefreshTime.getAndAdd(_refreshInterval) : 0;
	 }
}
```

should be changed to:

```
final class DefaultFaceletCache extends FaceletCache<DefaultFacelet> {

	long getAndUpdateNextRefreshTime() {
		// There is no point in calculating the next refresh time if we are refreshing always/never 		// There is no point in incrementing either, just SET it to expire from NOW! 		return (_refreshInterval > 0) ? _nextRefreshTime.getAndSet(System.currentTimeMillis() + _refreshInterval) : 0;	
	}
}
```

—

suppose 't0' is any instant on timeline:

```
**now: t0**create facelet record and put in cache
record._creationTime := t0
record._lastModified := t0 - 10,000ms (arbitrarily chosen in the past)
record._refreshInterval := 2,000ms (arbitrarily chosen)
record._nextRefreshTime := t0 + 2,000ms
```

then, suppose that this facelet is requested again after 1,000,000ms and has not been modified on filesystem:

```
**now: t0 + 1,000,000ms**get/read record from cache
record._creationTime = t0
record._lastModified = t0 - 10,000ms
record._refreshInterval = 2,000ms
record._nextRefreshTime = t0 + 2,000ms

check for expiration: 
    record._nextRefreshTime' < 'now' ==> 
    t0 + 2,000ms < t0 + 1,000,000ms ==> 
    expired!

update: record._nextRefreshTime := t0 + 2,000ms + 2,000ms = t0 + 4,000ms
check file modification: Util.getLastModified(url) (heavy operation!) ==> result: not modified
```

now suppose that record is accessed again and again, with 1ms interval between accesses (just point the first):

```
**now: t0 + 1,000,001ms**get/read record from cache
record._creationTime = t0
record._lastModified = t0 - 10,000ms
record._refreshInterval = 2,000ms
record._nextRefreshTime = t0 + 4,000ms

check for expiration: 
    record._nextRefreshTime < 'now' ==> 
    t0 + 4,000ms < t0 + 1,000,001ms ==> 
    expired! ==> 

    WRONG!
```

The record is seen as expired even if it has been accessed (and found expired - then 'refreshed') 1ms ago.

This cause <tt>Util.getLastModified(url)</tt> (**heavy op**) to be called on each subsequent access (including this access).

And this will continue on, until the access threshold is included within ('now'; 'now' + '_refreshInterval'] (depending on '_refreshInterval', this may also be impossible?).

So, this algo is not efficient. And it's because there's no conceptual reason to 'increment' the threshold.
Instead, just resetting the threshold on expire (and expunging the record if modified) is sufficient to maintain the algo safe and make it more efficient.

The concepual constraint, in this case, is: 'let x be a record: there is at most one expiration in any possible refresh interval, for record x'.
In other words: if record x has expired less than '_refreshInterval' ms ago, it's useless to check it again.

Although there's no formal proof of correctness, I don't see any side effect in resetting the threshold - no invalid case comes in my mind - maybe your eyes can spot something wrong yet.

If you are concerned about testing, I'll be glad to help you as much as I can.

Thank you, and keep up with the good work ![](https://java.net/jira/images/icons/emoticons/wink.gif)
#### Environment
no environment needed, the issue is conceptually evident.
#### Affected Versions
[2.3.0-m06]
